File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, n = s.size(), maxLen = 0;
        unordered_map<char,int> map;
        for(int right=0;right<n;right++){
            char curr = s[right];
            map[curr]++;

            while(map[curr] > 1){
                map[s[left]]--;
                left++;
            }
            maxLen = max(maxLen, right - left + 1);
        }
        return maxLen;
        
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0, maxLen = 0, n = s.length();
        Map<Character, Integer> map = new HashMap<>();

        for(int right=0;right < n ;right++){
            // add element 
            char curr = s.charAt(right);

            if(map.containsKey(curr)){// ccurr already present 
                left = Math.max(map.get(curr)+1, left);
            }
            map.put(curr, right);
            maxLen = Math.max(maxLen, right - left + 1);

        }
        return maxLen;        
    }
}